### ISSUE TITLE
TimeCode from Badly Formated SRT

### ISSUE THREAD

**Issue by X-Raym** [SubtitleEdit/subtitleedit#2191]
Hi !

Some SRT generator platforms are not really strict with the timecode format,
essentially, they often miss some digits.

These badly formated timecodes can read by Subtitle Edit, but it messes with timing.

For eg, with [DownSub](https://downsub.com/) you can get subs like this
```
1
00:00:09,9 --> 00:00:11,65
Bonjour à tous.
```

But once imported with SE, it becomes
```
1
00:00:09,009 --> 00:00:11,065
Bonjour à tous.
```
![Screnshot](https://monosnap.com/file/QadUOdUqaFjsqEHPzoadzLYlzEaT9e.png)

So it goes from 9.3 to 9.003, which can create sync error to a maximum of 881ms, which is pretty huge. These timing issues are committed to the file at save.

Thanks for your support, and for your great software!

**Comment by niksedk** [comment]
OK, it's probably more likely that "00:00:09,9" should have been "00:00:09,900" than "00:00:09,009", but impossible to know for sure - but it should work better for you in next update.

**Comment by X-Raym** [comment]
@niksedk 
Thanks for your quick answer ! :)

Sorry for the delay, I had to find a tool that would allow me to download proper SRT from YouTube to double check this issue, and it took some times. I should have done that before but I didn't find the right tool at this time (most online downloader doesn't even care about milliseconds).
I finally find a tool which display timecodes from youtube accurately: YouTube Subtitle Download addons FireFox, which is more precise and strict with formatting. Here are the results:

* `00:00:09,9 --> 00:00:11,65` for DownSub
* `00:00:09,009 --> 00:00:11,065` for DownSub open with SE.
* `00:00:09,009 --> 00:00:11,065` for YouTube Subtitle Download (which seems accurate)

So in fact... I think the actual behavior already fit the desire output... which is quite inuintive considering how the data are formatted. So yes, DownSub format, with `,` + `number of milliseconds` seems to be already well interpreted by SE. :/

I get confused by DownSub formatting... I think this issue can be considered as a false report... but I feel a bit guilty, as you already committed code fixes :/

**Comment by X-Raym** [comment]
That said, DownSubs timecode format maybe the exception, so maybe we could have a popup to let us choose how to globally interpret that kind of incomplete milliseconds format.

What do you think ?

**Comment by aaaxx** [comment]
@X-Raym, can you link to the YouTube video?

**Comment by X-Raym** [comment]
@aaaxx 

- here is a YouTube [video](https://www.youtube.com/watch?v=5MgBikgcWnY) with subtitles.
- Here is [DownSub](http://downsub.com/), the source of the confusion.
- and here is [YouTube Subtitle Downloader](https://addons.mozilla.org/fr/firefox/addon/youtube-subtitle-downloader/), which has properly formatted SRT.

**Comment by aaaxx** [comment]
OK, here's how YouTube actually stores them: [http://video.google.com/timedtext?lang=fr&v=5MgBikgcWnY](http://video.google.com/timedtext?lang=fr&v=5MgBikgcWnY). The format is unpadded seconds and _padded_ decimal remainder. The first line starts at 9.003 and has the duration of 2.062, which makes the end time 11.065.

Also, TED Talks use [Amara](http://amara.org) (online translation platform) for subtitling their videos, so you can get all of their source files there. The SRT for [this video](http://amara.org/en/videos/r6efi6FlN3Ef/fr/553688/) uses valid timecodes and the first line is `00:00:09,003 --> 00:00:11,065`, which matches the one stored by YouTube.

(By the way, both your downloaders give me the number 3, not 9, in in the start timecode of the French file.)

So, clearly, DownSub is the odd one out and is seriously broken:

- `00:00:09,3 --> 00:00:11,65` DownSub
- `00:00:09,003 --> 00:00:11,065` Youtube Subtitle Downloader
- `00:00:09,003 --> 00:00:11,065` <b>Amara (original)</b>
- `00:00:09,003 --> 00:00:11,065` Youtube (stores it as 9.003 + 2.062 duration)

**Comment by aaaxx** [comment]
And, @niksedk, regardless of this specific case, here's how I see it. The only logical thing to do when reading files with non-standard timecodes (i.e. non-zero-padded when the particular subtitle format expects them to be) is to pad the integer units (frames, hours, minutes, seconds) with *leading* zeros and decimal units (e.g. milliseconds in SRT) with *trailing* zeros, because that's the only way to change their formatting without changing their value.

In other words, it doesn't make sense to assume that the writing application did anything other than simply wrote correct values without zero padding them, because any other behaviour (such as opening a correctly formatted file and removing the padding from the wrong side, then later saving it) would be a bug.

**Comment by X-Raym** [comment]
@aaaxx 
That make sense.
Actually when we saved DownSub subtitles from SE, it add the leading zeros, so it works good as expected, contrary to my initial report.

I think we can close this issue for now then.

(ps : the timecode extract I gave as example had 9 instead of 3 cause I wanted to highlight the maxium time difference it could have, just a simple decimal replacement).

**Comment by MaximPro** [comment]
Here is a solution anyway, if it's still needed.
Download & Use subtitle text editor, open srt file, use multiple cursors feature and choose all lines at once (CTRL+A  +  CTRL+L) and fix all zeros at once manually! takes only a few seconds ;)
